### ISSUE TITLE
nan when using np.where and grad

### ISSUE THREAD

**Issue by mrtupek** [google/jax#5798]
Here is the situation:

```
f = lambda x: np.where(x<1., x, np.sqrt(x))
df = grad(f)
df(1e-9)  # evaluates to 1, as expected
df(0.0)   # evaluates to nan, presumably because the second case of the where is nan in that case
         # but it is the first case that should give the result, not the second!
```

Not sure if this is expected behavior, but it is surprising.  I'm going to work on a work-around, but thought I should document it.  Thanks!

**Comment by jakevdp** [comment]
Hi – thanks for the clear question! This is a well-known "feature" of autograd frameworks, that is generally solved by using the "double where trick" (see https://github.com/google/jax/issues/1052#issuecomment-514083352 for a full explanation)

For your example, the application of this solution might look something like this:
```lang-python
import jax.numpy as jnp
from jax import grad

def f(x):
  condition = x < 1
  x_valid = jnp.where(condition, 1, x)
  return jnp.where(condition, x, jnp.sqrt(x_valid))

df = grad(f)
print(df(1e-9))
# 1.0
print(df(0.0))
# 1.0
```

**Comment by mattjj** [comment]
I think this is also [mentioned in the FAQ](https://jax.readthedocs.io/en/latest/faq.html#gradients-contain-nan-where-using-where).

**Comment by mrtupek** [comment]
Thanks for the clear solution!  Sorry I didn't find the FAQ earlier.  For what it is worth, here is an alternative solution I found (redefining sqrt to be 'safe' when taking grads):

```
import jax.numpy as np
from jax import grad, lax, custom_jvp

@custom_jvp
def safe_sqrt(x):
    return np.sqrt(x)


@safe_sqrt.defjvp
def safe_sqrt_jvp(xpack, vpack):
    x, = xpack
    v, = vpack
    f = safe_sqrt(x)
    df = v * lax.cond( x <= 0,
                       lambda x: 0.,
                       lambda x: 0.5/np.sqrt(x),
                       x )
    return f, df


f = lambda x: np.where(x<1., x, safe_sqrt(x))
df = grad(f)

print(df(1e-9))
print(df(0.0)) 
```

Does this also seem like a reasonable solution?  Now whenever I call need a sqrt in grad'd code I can just use this 'safe' version to avoid the double where in my app code?  Is there an obvious gotcha here?

**Comment by mattjj** [comment]
There is something of a gotcha, but it's not obvious! It's related to the one mentioned in [that comment on #1052](https://github.com/google/jax/issues/1052#issuecomment-514083352) that Jake linked.

```python
print(grad(lambda x: safe_sqrt(x)**2)(0.))  # prints 0.0
```

I think that should be `nan` because the function isn't differentiable over the reals at 0. It is differentiable over the nonnegative reals, but there the derivative is 1.0, not 0.0 as computed. (Related: [enforcing differentiation conventions using `custom_jvp`](https://jax.readthedocs.io/en/latest/notebooks/Custom_derivative_rules_for_Python_code.html#enforcing-a-differentiation-convention).)

**Comment by mrtupek** [comment]
Interesting... I see the dilemma now.  

I think for my usage, I have to conclude that the wrong derivative of sqrt(x)^2 at x=0 is probably the lesser of two evils.  Especially because my actual use case is to construct a 'smooth' sqrt(x dot x) which is quadratic around x=0, so I'll never actually evaluate it there.  Thanks a lot for the help and the links.
